pingcap/tiflash
# Restore checkpoint location after restart

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/7400

Problem Summary:

The checkpoint info could possible lag behind the S3 manifest after restart. If we only rely on the local persisted WAL to restore the checkpoint info into memory, it will cause data inconsistency problem that a tiflash wn referring a deleted S3 CheckpointDataFile.

### What is changed and how it works?

* For a bootstrapped write node, wait for fetching checkpoint info from S3 before receiving GRPC requests
* After restarting, try to fetch the checkpoint info from the latest manifest on S3, for each page_id in the manifest:
  * If a local entry has the same sequence as in the S3 manifest, then copy the checkpoint info
  * If a local entry has newer sequence to the S3 manifest, then the page is updated, just ignore
  * If the page_id does not exist in local after restart, it means the page_id is deleted, just ignore
* Add an HTTP API "http://127.0.0.1:20170/tiflash/sync-remote-store" to re-sync all local page data to S3

### Check List

Tests 

  - Deploy a disaggregated cluster with 1 wn, 1 cn
  - Enable `profiles.default.remote_gc_verify_consistency: 1`
  - Run chbenchmark and check whether all check passed

Side effects


Documentation


### Release note



```release-note
None
```
